codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -0,0 +1,297 @@
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPromise;
+import io.netty.channel.EventLoopGroup;
+import io.netty.handler.codec.http.HttpServerUpgradeHandler.UpgradeEvent;
+import io.netty.util.concurrent.EventExecutor;
+import io.netty.util.internal.OneTimeTask;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class Http2MultiplexCodec extends ChannelHandlerAppender {
+    private static final Http2FrameLogger HTTP2_FRAME_LOGGER
+        = new Http2FrameLogger(INFO, Http2MultiplexCodec.class);
+
+    private final ChannelHandler streamHandler;
+    private final EventLoopGroup streamGroup;
+    private final Http2ConnectionHandler http2Handler;
+    private ChannelHandlerContext ctx;
+    private final List<StreamInfo> streamsToFireChildReadComplete = new ArrayList<StreamInfo>();
+
+    /** {@code streamHandler} should be {@link ChannelHandler.Sharable}. */
+    public Http2MultiplexCodec(boolean server, ChannelHandler streamHandler) {
+        this(server, streamHandler, null);
+    }
+
+    /** {@code streamHandler} should be {@link ChannelHandler.Sharable}. */
+    public Http2MultiplexCodec(boolean server, ChannelHandler streamHandler,
+            EventLoopGroup streamGroup) {
+        super(false /* selfRemoval */);
+        this.streamHandler = streamHandler;
+        this.streamGroup = streamGroup;
+        http2Handler = new Http2ConnectionHandler(new DefaultHttp2Connection(server),
+                new Http2InboundFrameLogger(new DefaultHttp2FrameReader(), HTTP2_FRAME_LOGGER),
+                new Http2OutboundFrameLogger(new DefaultHttp2FrameWriter(), HTTP2_FRAME_LOGGER),
+                new FrameListener());
+        http2Handler.connection().addListener(new ConnectionListener());
+        add(http2Handler);
+    }
+
+    Http2ConnectionHandler connectionHandler() {
+        return http2Handler;
+    }
+
+    @Override
+    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+        this.ctx = ctx;
+        super.handlerAdded(ctx);
+    }
+
+    /**
+     * Handles the cleartext HTTP upgrade event. If an upgrade occurred, sends a simple response via
+     * HTTP/2 on stream 1 (the stream specifically reserved for cleartext HTTP upgrade).
+     */
+    @Override
+    public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
+        if (!(evt instanceof UpgradeEvent)) {
+            super.userEventTriggered(ctx, evt);
+        }
+
+        UpgradeEvent upgrade = (UpgradeEvent) evt;
+        Http2Stream stream = http2Handler.connection().requireStream(1);
+        // The stream was already made active, but ctx may have been null so it wasn't initialized.
+        new ConnectionListener().onStreamActive(stream);
+        upgrade.upgradeRequest().headers().setInt(HttpUtil.ExtensionHeaderNames.STREAM_ID.text(), 1);
+        new InboundHttpToHttp2Adapter(http2Handler.connection(), new FrameListener())
```

Comment:
Should we be using the same `FrameListener` that we allocated in the constructor?